Update Symfony documentation (#923)
brefphp/bref: docs/frameworks/symfony.md

This guide helps you run Symfony applications on AWS Lambda using Bref. These instructions are kept up-to-date to be compatible with the latest Symfony version.

Multiple demo applications are available on GitHub at [github.com/brefphp/examples/Symfony](https://github.com/brefphp/examples/tree/master/Symfony).

## Setup

First, make sure you have followed the [Installation guide](../installation.md) to create an AWS account and install the necessary tools.

Next, in an existing Symfony project, install Bref and the [Symfony Bridge package](https://github.com/brefphp/symfony-bridge).

```
composer require bref/symfony-bridge
```

> If you are using [Symfony Flex](https://flex.symfony.com/),it will automatically run

Reviewer: missing space :smile: 

```suggestion
If you are using [Symfony Flex](https://flex.symfony.com/), it will automatically run
```

This guide helps you run Symfony applications on AWS Lambda using Bref. These instructions are kept up-to-date to be compatible with the latest Symfony version.

Multiple demo applications are available on GitHub at [github.com/brefphp/examples/Symfony](https://github.com/brefphp/examples/tree/master/Symfony).

## Setup

First, make sure you have followed the [Installation guide](../installation.md) to create an AWS account and install the necessary tools.

Next, in an existing Symfony project, install Bref and the [Symfony Bridge package](https://github.com/brefphp/symfony-bridge).

```
composer require bref/symfony-bridge
```

If you are using [Symfony Flex](https://flex.symfony.com/), you're done. Flex will automatically run
the [bref/symfony-bridge recipe](https://github.com/symfony/recipes-contrib/tree/master/bref/symfony-bridge/0.1) which will perform the following tasks:

- Create a `serverless.yml` configuration file optimized for Symfony.
- Add the `.serverless` folder to the `.gitignore` file.

> Otherwise, you can create the `serverless.yml` file manually at the root of the project, with the following configuration:

Reviewer: Just an idea : Maybe we could link to the recipes repositiory for those not using Flex, that would avoid synching the recipe and the docs every time we need to do a change ?

Most projects will use Flex as choosing Bref is rarely something you do on a legacy/not updated project.

WDYT ?

Author: Indeed, this will be easier to maintain afterwards.

Author: I removed the configuration example and added an annotation to recipe with a direct link

service: symfony

provider:
  name: aws
  # The AWS region in which to deploy (us-east-1 is the default)
  region: us-east-1
  # The stage of the application, e.g. dev, production, staging… ('dev' is the default)
  stage: dev
  runtime: provided.al2
  environment:
    # Symfony environment variables
    APP_ENV: prod

plugins:
  - ./vendor/bref/bref

functions:
  # This function runs the Symfony website/API
  web:
    handler: public/index.php
    timeout: 28 # in seconds (API Gateway has a timeout of 29 seconds)
    layers:
      - ${bref:layer.php-80-fpm}
    events:
      - httpApi: '*'
  # This function let us run console commands in Lambda
  console:
    handler: bin/console
    timeout: 120 # in seconds
    layers:
      - ${bref:layer.php-80} # PHP
      - ${bref:layer.console} # The "console" layer

package:
  patterns:
    # Excluded files and folders for deployment
    - '!node_modules/**'
    - '!tests/**'
    - '!var/**'
    # If you want to include files and folders that are part of excluded folders,
    # add them at the end
    - 'var/cache/prod/**'
```

You still have a few modifications to do on the application to make it compatible with AWS Lambda.

> Since [the filesystem is readonly](/docs/environment/storage.md) except for `/tmp` you need to customize where the cache and logs are stored in
> the `src/Kernel.php` file. This is done by extending the default `Kernel` class create by Symfony with the `Kernel` class provide by the bridge:

Reviewer: ```suggestion
Since [the filesystem is readonly](/docs/environment/storage.md) except for `/tmp` we need to customize where the cache and logs are stored 
in the `src/Kernel.php` file. This is done automatically by the bridge, you just need to use the `BrefKernel` class provided by the bridge instead of the default `BaseKernel`
```

```diff
// src/Kernel.php

namespace App;

+ use Bref\SymfonyBridge\BrefKernel;
use Symfony\Bundle\FrameworkBundle\Kernel\MicroKernelTrait;
use Symfony\Component\Config\Loader\LoaderInterface;
use Symfony\Component\Config\Resource\FileResource;
use Symfony\Component\DependencyInjection\ContainerBuilder;
-use Symfony\Component\HttpKernel\Kernel as BaseKernel;
use Symfony\Component\Routing\RouteCollectionBuilder;

- class Kernel extends BaseKernel
+ class Kernel extends BrefKernel
{
    // ...
```

> You can also separate the [build directory](https://symfony.com/doc/current/reference/configuration/kernel.html#build-directory)

Reviewer: I've tested this several times and this works really bad with the bridge for now (see for example https://github.com/brefphp/symfony-bridge/pull/36 and https://github.com/symfony/symfony/issues/40827)

I would delete this for now as it causes more problems than it solves and the bridge is more performant on projects where I tested both approaches.

Author: Fixed

from the cache directory to be able to [deploy a pre-warmed cache](#Deploy) and thus improve load time.

```php
// src/Kernel.php

class Kernel extends BrefKernel
{
    // ...
    

>     public function getBuildDir(): string

Reviewer: and remove this too

Author: Also fixed

This guide helps you run Symfony applications on AWS Lambda using Bref. These instructions are kept up-to-date to be compatible with the latest Symfony version.

Multiple demo applications are available on GitHub at [github.com/brefphp/examples/Symfony](https://github.com/brefphp/examples/tree/master/Symfony).

## Setup

First, make sure you have followed the [Installation guide](../installation.md) to create an AWS account and install the necessary tools.

Next, in an existing Symfony project, install Bref and the [Symfony Bridge package](https://github.com/brefphp/symfony-bridge).

```
composer require bref/symfony-bridge
```

If you are using [Symfony Flex](https://flex.symfony.com/), you're done. Flex will automatically run
the [bref/symfony-bridge recipe](https://github.com/symfony/recipes-contrib/tree/master/bref/symfony-bridge/0.1) which will perform the following tasks:

- Create a `serverless.yml` configuration file optimized for Symfony.
- Add the `.serverless` folder to the `.gitignore` file.

> Otherwise, you can create the `serverless.yml` file manually at the root of the project. Take a look

> at the [default configuration](https://github.com/symfony/recipes-contrib/blob/master/bref/symfony-bridge/0.1/serverless.yaml) provide by the recipe.

Reviewer: ```suggestion
at the [default configuration](https://github.com/symfony/recipes-contrib/blob/master/bref/symfony-bridge/0.1/serverless.yaml) provided by the recipe.
```

## Trust API Gateway

When hosting your site on Lambda, API Gateway will act as a proxy between the client and your function.

By default, Symfony doesn't trust proxies for security reasons, but it's safe to do it when using API Gateway and Lambda.

This is needed because otherwise, Symfony will not be able to generate URLs properly.

You should add the following lines to `config/packages/framework.yaml`

```yaml
# config/packages/framework.yaml

framework:
  # trust the remote address because API Gateway has no fixed IP or CIDR range that we can target

>   trusted_proxies: '127.0.0.1,REMOTE_ADDR'

Reviewer 2: Isn't `$_SERVER['REMOTE_ADDR']` always 127.0.0.1 on Lambda?

Reviewer: I think it is https://github.com/brefphp/bref/blob/master/src/Event/Http/FpmHandler.php#L184

So we could change it to only keep the IP

```suggestion
  trusted_proxies: '127.0.0.1'
```

Reviewer 2: :+1:

  # trust "X-Forwarded-*" headers coming from API Gateway
  trusted_headers: [ 'x-forwarded-for', 'x-forwarded-proto', 'x-forwarded-port' ]
```

Note that API Gateway doesn't set the `X-Forwarded-Host` header, so we don't trust it by default. You should only whitelist this header if you set it manually,
for example in your CloudFront configuration (see how to do it in
the [example Cloudformation template](../websites.md#serving-php-and-static-files-via-cloudfront)).

> Be careful with these settings if your app will not be executed only in a Lambda environment.

You can get more details in the [Symfony documentation](https://symfony.com/doc/current/deployment/proxies.html).

### Getting the user IP

**When using CloudFront** on top of API Gateway, you will not be able to retrieve the client IP address, and you will instead get one of Cloudfront's IP when
calling `Request::getClientIp()`. If you really need this, you will need to

> whitelist [every CloudFront IP](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/LocationsOfEdgeServers.html)

Reviewer 2: I know you did not change this. But this is not really true. 

You can just update your index.php file with: 

```php
// Get user IP:
if (isset($_SERVER['LAMBDA_REQUEST_CONTEXT'])) {
    $context = json_decode($_SERVER['LAMBDA_REQUEST_CONTEXT'], true);
    $_SERVER['HTTP_X_FORWARDED_FOR'] = $context['identity']['sourceIp'] ?? '';
}
```

Reviewer: There was some discussions about this on Slack and it seems to work fine but we weren't able to say if this is working in all scenarios. And I wasn't able to find AWS docs that states clearly what they put in it.

If you are sure about that, I'm a :100: to document it :+1:

in `trusted_proxies`.

## Assets

To deploy Symfony websites, assets needs to be served by AWS S3. Setting up the S3 bucket is already explained in
the [websites documentation](../websites.md#hosting-static-files-with-s3). This section provides additional instructions specific to Symfony assets and Webpack
Encore.

First, you need to [tell Symfony](https://symfony.com/doc/current/reference/configuration/framework.html#base-urls) to use the S3 URL as the assets base URL,
instead of your app domain in production.

```yaml
# config/packages/prod/assets.yaml

framework:
  assets:

>     base_urls: 'https://<bucket-name>.s3.amazonaws.com'

Reviewer 2: Should we really care about S3? Why not do cloudfront directly?

Reviewer: This is simpler to setup for beginners and if you put Cloudfront in front of your app, then this is not needed I think.

## Console

As you may have noticed, we define a function of type "console" in `serverless.yml`. That function is using the [Console runtime](/docs/runtimes/console.md),
which lets us run the Symfony Console on AWS Lambda.

To use it follow [the "Console" guide](/docs/runtimes/console.md).

## Logs

> While overriding the log's location in the `Kernel` class was necessary for Symfony to run correctly, by default Symfony logs in `stderr`. That is great because

Reviewer: We don't need to say this as it is internal to the Bridge

```suggestion
By default Symfony logs in `stderr`. That is great because
```

Author: Indeed, I changed it

If using Webpack Encore, you also need to add the following config at the end of `webpack.config.js`

```js
// webpack.config.js

if (Encore.isProduction()) {
    // Note the '/build' at the end of the URL
    Encore.setPublicPath('https://<bucket-name>.s3.amazonaws.com/build');
    Encore.setManifestKeyPrefix('build/')
}
```

Then, you can compile assets for production in the `public` directory, and synchronize that directory to a S3 bucket:

```bash
php bin/console assets:install --env prod
# if using Webpack Encore, additionally run
yarn encore production
aws s3 sync public/ s3://<bucket-name>/ \

>   --delete \

Reviewer 2: If you delete stuff, it would mean that there is some downtime for your application. (ie no assets)

Reviewer: True, I think the solution is to sync twice. First without `--delete` and then with the `--delete` option.

Or do you have another solution ?

Reviewer 2: I'll just leave the old files :)

  --exclude index.php \
  --exclude public/build/manifest.json \

>   --exclude public/build/entrypoint.json

Reviewer 2: Also, one should configure cace control (But only if one use unique filenames, ie hashes)

```
--cache-control max-age=31449600,immutable
```

Reviewer: File hasing is enabled by default with Webpack Encore. But I'm not sure it's true for third party bundles providing assets (like EasyAdmin or Sonata for example) moved by `bin/console assets:install` in `bundles`.

```

Finally, you need to update the `serverless.yml` file to exclude the `assets`, `public/build` and `public/bundles` directories from deployment:

```yaml
# serverless.yml

package:

>   pattern:

Reviewer: it is patterns with an `s`

Author: it's fixed, thanks @t-richard

If using Webpack Encore, you also need to add the following config at the end of `webpack.config.js`

```js
// webpack.config.js

if (Encore.isProduction()) {
    // Note the '/build' at the end of the URL
    Encore.setPublicPath('https://<bucket-name>.s3.amazonaws.com/build');
    Encore.setManifestKeyPrefix('build/')
}
```

Then, you can compile assets for production in the `public` directory, and synchronize that directory to a S3 bucket:

```bash
php bin/console assets:install --env prod
# if using Webpack Encore, additionally run
yarn encore production
aws s3 sync public/ s3://<bucket-name>/ \
  --delete \
  --exclude index.php \
  --exclude public/build/manifest.json \
  --exclude public/build/entrypoint.json
```

Finally, you need to update the `serverless.yml` file to exclude the `assets` and `public/build` directories from deployment:

```yaml
# serverless.yml

package:
  pattern:
    - '!assets/**'
    - '!public/build/**'

>     - 'public/build/manifest.json'

Reviewer: maybe add `!public/bundles` that may be generated by `bin/console assets:install`

    - 'public/build/entrypoint.json'
```

### Assets in templates

For the above configuration to work, assets must be referenced in templates via the `asset()` helper:

```diff

> - <img src="/images/logo.png">

Reviewer: Shouldn't it be the following ?

```suggestion
- <img src="/build/images/logo.png">
```

Author: After thought, I think we can totally remove this part because Symfony recommends to use the Asset component to display images anyway: https://symfony.com/doc/current/templates.html#linking-to-css-javascript-and-image-assets

Reviewer: Well there are two ways to see this :
1. this is common knowledge that you should use `asset()` so no need to document it
2. some users might not know this and this small bit of docs will help them and avoid support from the Bref community for these users 

From my personal experience, there are some users that don't know about this and just use absolute paths in their templates.

So I would keep it.

It's also the same in the laravel docs https://bref.sh/docs/frameworks/laravel.html#assets-in-templates

But maybe linking to that Symfony doc for details about this could be a good idea.

Author: It makes sense and remains consistent with the Laravel doc indeed, PR updated 😉
